README.md
[![Travis-CI Build Status](https://travis-ci.org/norival/theiaR.svg?branch=master)](https://travis-ci.org/norival/theiaR)
[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/norival/theiaR?branch=master&svg=true)](https://ci.appveyor.com/project/norival/theiaR)
[![CRAN status](https://www.r-pkg.org/badges/version-ago/theiaR)](https://cran.r-project.org/package=theiaR)
![GitHub R package version](https://img.shields.io/github/r-package/v/norival/theiaR)
![GitHub R package version (devel)](https://img.shields.io/github/r-package/v/norival/theiaR/devel)
[![downloads](https://cranlogs.r-pkg.org/badges/grand-total/theiaR)](https://cran.r-project.org/package=theiaR)
[![lifecycle](https://img.shields.io/badge/lifecycle-stable-green.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)

# TheiaR: search, download and manage data from Theia

The TheiaR package provides an efficient and clean interface to search, download
and manage products from [Theia website](https://theia.cnes.fr/atdistrib/rocket/#/home).


## Overview

The basic functionalities are:

- Search available tiles through Theia API
- Download tiles resulting from a search
- Download tiles contained in a cart (`.meta4` file) obtained from Theia
  website.
- Read band into `RasterStack` objects (with the `raster` library)
- Read band into `gdalcubes` objects (with the `gdalcubes` library)
- Extract archives (not recommended)

__NOTE:__ To search and download data from Theia, you will need to [register to
their website](https://sso.theia-land.fr/theia/register/register.xhtml).

__NOTE:__ In order to use _Landsat_ or _SpotWorldHeritage_ products, you'll need
to make a first manual download to agree to the license and validate your
account.


## Installation

You can install the latest release from CRAN:

``` r
isntall.packages('theiaR')
```

Alternatively, you can install the latest stable version from Github with:

``` r
devtools::install_github('norival/theiaR')
```

Or the latest development version:

``` r
devtools::install_github('norival/theiaR', 'devel')
```


## Complete example

A workflow to search and download tiles would be something like:

``` r
library(theiaR)

# create a list containing the query
myquery <- list(collection = "SENTINEL2",
                town       = "Grenoble",
                start.date = "2018-07-01",
                end.date   = "2018-07-06")

# create a collection from the query
mycollection <- TheiaCollection$new(query = myquery, dir.path = ".", check = TRUE)

# check available tiles from the query
mycollection$status

# download the tiles into 'dir.path'
mycollection$download(auth = "path/to/auth/file.txt")
```


## Step-by-step guide


### Create a collection of tiles

The first step is to create a collection of tile(s). This can be done either
from a query or from a cart file (downloaded from Theia's website).


#### Create a collection from a query

A query is simply a named `list` of search terms. For example:

``` r
myquery <- list(collection = "SENTINEL2",
                town       = "Grenoble",
                start.date = "2018-07-01",
                end.date   = "2018-07-06")
```

will create a query to Theia database, looking for tiles from Sentinel2
satellite around Grenoble, between 2018-07-01 and 2018-07-31.

See the
[vignette](https://cran.r-project.org/package=theiaR/vignettes/theiaR.html#create-a-collection-from-a-query)
for all the available options.

You can then create your collection with:

``` r
mycollection <- TheiaCollection$new(query = myquery, dir.path = ".", check = TRUE)
```

where `dir.path` is the path you want your tiles to be further downloaded (This
only queries Theia's catalog for available tiles, nothing is downloaded). If
tiles are already present in `dir.path`, they will be checked by computing a
checksum and comparing it to the hash provided by Theia (only available for
Sentinel2 data, no hash is provided for other collections, and files are then
assumed to be correct). This ensures that the files have been correctly
downloaded. Set `check = FALSE` to skip file's check.


#### Create a collection from a cart file

Alternatively, you can download a cart from Theia. To create a cart, login to
Theia website, make a [search](https://theia.cnes.fr/atdistrib/rocket/#/home)
for tiles, and add wanted tiles to your cart. Then, download your cart and save
the resulting `.meta4` file to your disk.

You can then create your collection using this file:

``` r
mycollection <- TheiaCollection$new(cart.path = "path/to/cart/file.meta4",
                                    dir.path  = ".",
                                    check     = TRUE)
```

As above, it will check the hash of files if they are already present in
`dir.path`.


### Getting information on your collection

You can access the tiles from your collection using:

```
mycollection$tiles
```

which returns a `list` of tiles. You can also see the status of your collection
with:

``` r
mycollection$status
```


### Download your tiles

The next step is to download your collection. To download all tiles in a
collection, simply run:

``` r
mycollection$download(auth = "path/to/auth/file.txt")
```

where `path/to/auth/file.txt` is the path to a file storing your Theia
credentials. It is a simple text file with the Theia's account email on the
first line and the account's password on the second line:

``` txt
user@example.com
MyTheiaPassword
```

If it does not exist yet, you will be securely prompted for your login and
password, and the file will be created.

The `download()` method will check if files are present, check their hashes,
and download them if needed (if files do not exist or checksums are wrong). To
overwrite existing files, run:

``` r
mycollection$download(auth = "path/to/auth/file.txt", overwrite = TRUE)
```


### Read bands from zip files

You can then read bands directly from the zip archives (by using the `vsizip`
interface provided by GDAL). Use:

``` r
mytile$bands
```

to get a list of available bands. Then:

``` r
mybands <- mytile$read(bands = c("B5", "B6"))
```

to load the bands into memory (returns a `RasterStack` object). It performs the
necessary corrections on the values.

You can also read bands from a collection by running:

``` r
mybands <- mycollection$read(bands = c("B5", "B6"))
```

which returns a `list` of `RasterStack` objects.

_NOTE: Be careful when loading several tiles as it needs a lot of memory (~900MB/tile)_


### Create a `gdalcubes` collection

Alternatively, you can use the great [gdalcubes](https://github.com/appelmar/gdalcubes_R)
package to create a three dimensional representation of the tiles. Simply run:

``` r
library(gdalcubes)

gdalcubes <- mycollection$as_gdalcube("path/to/gdalcubes.sqlite")
```

where `path/to/gdalcubes.sqlite` is the path to store the gdalcubes object data.


### Extract tiles

If you want to extract full archives, you can run:

``` r
file.path <- mycollection$extract()
```

which will extract tiles into the same directory as the archives.

**This is not recommended, as this will take a large amount of disk space**


## Acknowledgments

Thanks to Olivier Hagolle for his work on `theia_download.py`
([github](https://github.com/olivierhagolle/theia_download)), which has inspired
this package.


NEWS.md
# theiaR 0.4.3

## Minor changes

- Add "VENUSVM05" collection

# theiaR 0.4.2

## Minor changes

- Allow "tile" field in queries for all collections except VENUS

## Bug fixes

- Fix processing level names for Snow collection [Issue #5](https://github.com/norival/theiaR/issues/5#issuecomment-1026136014)
- Fix cran checks note

# theiaR 0.4.1

## Bug fixes

- Fix bad default processing level for some collections
- Fix cran checks note


# theiaR 0.4.0

## Minor changes

- Adds error when end.date is lower or equal to start.date

## Bug fixes

- Fix query metadata reading for Landsat57 format
- Fix query URL and level compatibility for Landsat57 collection
- Update README, vignette and documentation
- Fix error for new landsat metadata format


# theiaR 0.3.0

## New features

- Adds export to gdalcubes' image collection

## Bug fixes

- Fix error due to bad file path on Windows

## Minor changes

- Adds quiet arguments to control verbose output
- Adds 'check' argument to 'TheiaTile' and 'TheiaCollection' download methods
- Do not parse meta.data on tile creation, but do it on the fly


# theiaR 0.2.1

## Bug fixes

- Fix printing of TheiaQuery objects


# theiaR 0.2.0

## New features

- Adds orbit number and relative orbit number to queries
- Adds max records to queries
- Adds reading of bands from zip files
- Adds no check option for creation of Tiles


# theiaR 0.1.2

## Bug fixes

- Fix status dataframe for TheiaCollection objects


# theiaR 0.1.1

## Minor changes

- Adds processing level search term to queries
- Adds default values for collection and level to queries

## Bug fixes

- Fix metadata file detection


# theiaR 0.1.0

## Major changes

- First package release


R/TheiaQuery.R
#' A query to the Theia website
#'
#' Generate an send a query to Theia web API to get and download tiles based on
#' input given by the user.
#'
#' @name TheiaQuery
#'
#' @section Usage:
#' \preformatted{
#'    q <- TheiaQuery$new(query)
#'
#'    q$update_token()
#'    q$submit()
#' }
#'
#' @section Arguments:
#'
#' \describe{
#'    \item{q:}{A \code{TheiaQuery} object}
#'    \item{query:}{\code{list}, the users' request, see `Queries` for
#'    more informations}
#' }
#'
#' @section Details:
#'    \code{TheiaQuery$new()} Create a new instance of the class, parse `query`
#'    list and submit the query to Theia to retrieve files catalog
#'
#'    \code{q$submit()} Submit the query to Theia and get a list of tiles
#'    corresponding to search criteria
#'
#' @section Queries:
#'
#'    Search criteria are given with a `list` accepting these fields:
#'    \itemize{
#'      \item{collection:} The collection to look for. Accepted values are:
#'        'SENTINEL2', 'LANDSAT', 'Landsat57', 'SpotWorldHeritage', 'Snow',
#'        'VENUS' and 'VENUSVM05'.
#'        Defaults to 'SENTINEL2'
#'      \item{platform:} The platform to look for. Accepted values are:
#'        'LANDSAT5', 'LANDSAT7', 'LANDSAT8', 'SPOT1', 'SPOT2', 'SPOT3',
#'        'SPOT4', 'SPOT5', 'SENTINEL2A', 'SENTINEL2B', 'L2B-SNOW', 'L3B-SNOW'
#'      \item{level:} Processing level. Accepted values are: 'LEVEL1C',
#'        'LEVEL2A', LEVEL3A', 'N2A'. Defaults to 'LEVEL2A' (or 'N2A' if
#'        querying Landsat57 collection).
#'      \item{town:} The location to look for. Give a common town name.
#'      \item{tile:} The tile identifier to retrieve.
#'      \item{start.date:} The first date to look for (format: YYYY-MM-DD).
#'      \item{end.date:} The last date to look for (format: YYYY-MM-DD). Must be
#'        after start.date. Defaults to today's date.
#'      \item{latitude:} The x coordinate of a point
#'      \item{longitude:} The y coordinate of a point
#'      \item{latmin:} The minimum latitude to search
#'      \item{latmax:} The maximum latitude to search
#'      \item{lonmin:} The minimum longitude to search
#'      \item{lonmax:} The maximum longitude to search
#'      \item{orbit.number:} The orbit number
#'      \item{rel.orbit.number:} The relative orbit number
#'      \item{max.clouds:} The maximum of cloud cover wanted (0-100)
#'      \item{max.records:} The maximum of tiles to search
#'    }
#'
#' @seealso
#'    \url{https://github.com/olivierhagolle/theia_download} for an alternative
#'    download method based on Python. Inspiration for this function.
#'
#' @examples
#'
#' \donttest{
#' # Create a query to Theia database, looking for tiles from Sentinel2
#' # satellite around Grenoble, between 2018-07-01 and 2018-07-06.
#'
#' query <- list(collection = "SENTINEL2",
#'               town       = "Grenoble",
#'               start.date = "2018-07-01",
#'               end.date   = "2018-07-06")
#' q <- TheiaQuery$new(query)
#'
#' # Show informations on found tiles
#' print(q$tiles)
#' }
#'
NULL


#' @export

TheiaQuery <-
  R6Class("TheiaQuery",
          # private ------------------------------------------------------------
          private =
            list(url        = NULL,
                 catalog    = NULL,
                 server.url = NULL,
                 resto      = NULL,

                 
                 check      = function()
                 {
                   .TheiaQuery_check(self, private)
                 }),

          # public -------------------------------------------------------------
          public =
            list(query    = NULL,
                 tiles    = NULL,

                 initialize = function(query)
                 {
                   .TheiaQuery_initialize(self, private, query)
                 },

                 print = function(...)
                 {
                   .TheiaQuery_print(self)
                 },

                 submit = function()
                 {
                   .TheiaQuery_submit(self, private)
                 })
          )


# Functions definitions --------------------------------------------------------

.TheiaQuery_print <- function(self)
{
  # Special method to print
  cat("A search query to Theia website\n\n")

  cat("Number of tiles:", nrow(self$tiles), "\n\n")

  cat("Search terms:\n")

  # get maximum length of search terms to format printing
  max.chars <- max(nchar(names(self$query))) + 1

  # print search terms
  lapply(1:length(self$query),
         function(x) {
           l <- nchar(names(self$query))[[x]]

           cat("   ", names(self$query)[[x]], rep(" ", max.chars - l), ": ",
               self$query[[x]], "\n",
               sep = "")

           return(invisible(NULL))
         })

  return(invisible(self))
}


.TheiaQuery_initialize <- function(self, private, query)
{
  self$query <- query

  # check query list
  private$check()

  q.link <- list()

  q.link[["q"]]                   <- self$query$town
  q.link[["location"]]            <- self$query$tile
  q.link[["platform"]]            <- self$query$platform
  q.link[["processingLevel"]]     <- self$query$level
  q.link[["startDate"]]           <- self$query$start.date
  q.link[["completionDate"]]      <- self$query$end.date
  q.link[["lat"]]                 <- self$query$latitude
  q.link[["lon"]]                 <- self$query$longitude
  q.link[["orbitNumber"]]         <- self$query$orbit.number
  q.link[["relativeOrbitNumber"]] <- self$query$rel.orbit.number
  q.link[["maxRecords"]]          <- self$query$max.records

  # search a rectangle
  if (all(c("latmin", "latmax", "lonmin", "lonmax") %in% names(self$query))) {
    q.link[["box"]] <- URLencode(paste(self$query$lonmin,
                                       self$query$latmin,
                                       self$query$lonmax,
                                       self$query$latmax,
                                       sep = ","),
                                 reserved = TRUE)
  }

  # get fixed parts of links
  private$server.url <- "https://theia.cnes.fr/atdistrib/"
  private$resto      <- "resto2/"

  # fill query fields
  # build query links
  query.link  <- paste(names(q.link), q.link, sep = "=", collapse = "&")
  private$url <- paste0(private$server.url,
                        private$resto,
                        "api/collections/",
                        self$query$collection,
                        "/search.json?",
                        query.link)

  # submit the query
  self$submit()

  if (!(is.list(self$tiles)) && self$tiles ==  "No tiles matching search criteria") {
    # exit function if no tiles are found, with an error
    stop("No tiles matching search criteria", call. = FALSE)
  }

  if (!(is.null(query$max.clouds))) {
    # remove tiles with too much clouds
    clouds.over <- self$tiles$cloud.cover > query$max.clouds

    if (sum(clouds.over) > 0) {
      if (sum(clouds.over) == nrow(self$tiles)) {
        # if every tile has too much clouds, throw an informative error
        stop("Every tile has too much clouds (min cloud level: ",
             min(self$tiles$cloud.cover), ")",
             call. = FALSE)
      }

      # show number of bad tiles
      message("Removing ", sum(clouds.over), " over ", nrow(self$tiles),
              " tiles because too much clouds")

      # remove bad tiles
      self$tiles <- self$tiles[!(clouds.over), ]
    }
  }

  return(invisible(self))
}


.TheiaQuery_check <- function(self, private)
{
  # available choices
  collection.choices <- c('LANDSAT', 'Landsat57', 'SpotWorldHeritage', 'SENTINEL2', 'Snow', 'VENUS', 'VENUSVM05')
  platform.choices   <- c('LANDSAT5', 'LANDSAT7', 'LANDSAT8', 'SPOT1', 'SPOT2',
                          'SPOT3', 'SPOT4', 'SPOT5', 'SENTINEL2A', 'SENTINEL2B',
                          'VENUS')
  level.choices      <- c('LEVEL1C', 'LEVEL2A', 'LEVEL3A', 'N2A', 'L2B-SNOW', 'L3B-SNOW')

  # check queries
  self$query$tile       <- parse_query(self$query$tile, "tile", "character")
  self$query$town       <- parse_query(self$query$town, "town", "character")
  self$query$collection <- parse_query(self$query$collection, "collection", "character",
                                       choices = collection.choices,
                                       default = "SENTINEL2")
  self$query$platform   <- parse_query(self$query$platform, "platform", "character",
                                       choices = platform.choices)
  self$query$level      <- parse_query(self$query$level, "level", "character",
                                       choices = level.choices)
  self$query$start.date <- parse_query(self$query$start.date, "date", "character")
  self$query$end.date   <- parse_query(self$query$end.date, "date", "character",
                                       default = format(Sys.time(), "%Y-%m-%d")
                                      )
  self$query$max.clouds <- parse_query(self$query$max.clouds, "max.clouds", "numeric",
                                       choices = 0:100)
  self$query$latitude   <- parse_query(self$query$latitude, "latitude", "numeric")
  self$query$longitude  <- parse_query(self$query$longitude, "longitude", "numeric")
  self$query$latmin     <- parse_query(self$query$latmin, "latmin", "numeric")
  self$query$latmax     <- parse_query(self$query$latmax, "latmax", "numeric")
  self$query$lonmin     <- parse_query(self$query$lonmin, "lonmin", "numeric")
  self$query$lonmax     <- parse_query(self$query$lonmax, "lonmax", "numeric")

  self$query$orbit.number     <- parse_query(self$query$orbit.number, "orbit.number", "numeric")
  self$query$rel.orbit.number <- parse_query(self$query$rel.orbit.number, "rel.orbit.number", "numeric")
  self$query$max.records      <- parse_query(self$query$max.records, "max.records", "numeric",
                                             default = 500)

  # check for incompatible queries
  # if (!(is.null(self$query$tile)) && self$query$collection == "VENUS") {
  #   stop("'Tile' is not available for VENUS collection",
  #        call. = FALSE)
  # }

  # check that start date is lower than end date
  if (as.Date(self$query$end.date) - as.Date(self$query$start.date) <= 0) {
    stop("Invalid query: date. Start date must be lower than end date")
  }

  # check if user has not specified both a point and a rectangle
  box.names <- c("latmin", "latmax", "lonmin", "lonmax")
  if (any(box.names %in% names(self$query))) {
    if (!(all(box.names %in% names(self$query)))) {
      stop("Specify each of ", box.names, " to have a rectangle",
           call. = FALSE)
    }

    if (any(c("latitude", "longitude") %in% names(self$query))) {
      stop("Specify a point or a rectangle, not both", call. = FALSE)
    }
  }

  return(invisible(self))
}


.TheiaQuery_submit <- function(self, private)
{
  # TODO: gestion des erreurs
  # make http request to get catalog
  req <- httr::GET(private$url)

  httr::stop_for_status(req, task = paste0("retrieve data, invalid query url:\n", private$url))

  # parse and save catalog
  private$catalog <- httr::content(req, as = "parsed")

  if (length(private$catalog$features) == 0) {
    # exit function if no tiles are found
    self$tiles <- "No tiles matching search criteria"

    return(invisible(self))
  }
  # print(private$catalog$features[[1]]$properties)

  # extract tiles
  cart <-
    lapply(private$catalog$features,
           function(x) {
             # get file extension
             file.ext <- ifelse(grepl("gzip", x$properties$services$download$mimeType),
                                ".tar.gz",
                                ".zip")

             # return important information
             data.frame(file.name   = paste0(x$properties$productIdentifier, file.ext),
                        tile.id     = ifelse(is.null(x$id), NA, x$id),
                        file.hash   = ifelse(
                          is.null(x$properties$services$download$checksum),
                          NA,
                          x$properties$services$download$checksum
                        ),
                        cloud.cover =
                        ifelse(
                          is.null(x$properties$cloudCover),
                          NA,
                          as.numeric(as.character(x$properties$cloudCover))
                          ),
                        snow.cover  = ifelse(
                          is.null(x$properties$snowCover),
                          NA,
                          as.numeric(as.character(x$properties$snowCover))
                        ))
           })
  self$tiles <- do.call(rbind, cart)

  # build url for downloading
  self$tiles$url <- paste0(private$server.url,
                           private$resto,
                           "collections/",
                           self$query$collection,
                           "/",
                           self$tiles$tile.id,
                           "/download")

  return(invisible(self))
}
